### PR TITLE
Enable support for MQTT stitcher in stirling

### PR DESCRIPTION
**Pull request by ChinmayaSharma-hue**
Summary: This PR adds the stitcher component of MQTT (v5), a newly added protocol.

Related issues: https://github.com/pixie-io/pixie/issues/341

Type of change: /kind feature

Test Plan: Added tests

### CODE REVIEW

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+#include "src/stirling/source_connectors/socket_tracer/protocols/common/interface.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mqtt/types.h"
+
+namespace px {
+namespace stirling {
+namespace protocols {
+namespace mqtt {
+
+/**
+ * StitchFrames is the entry point of the Cassandra Stitcher. It loops through the resp_frames,
```
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.h:35]:
```suggestion
 * StitchFrames is the entry point of the MQTT Stitcher. It loops through the resp_frames,
```

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+      uint8_t request_match_key =
+          (req_frame.control_packet_type << 4) | static_cast<uint8_t>(req_frame.dup) << 3 |
+          (req_frame.header_fields["qos"]) << 1 | static_cast<uint8_t>(req_frame.retain);
+      auto iter = MapRequestToResponse.find(request_match_key);
+      if (iter == MapRequestToResponse.end()) {
+        VLOG(1) << absl::Substitute("Could not find any responses for frame type = $0",
+                                    request_match_key);
+        continue;
+      }
+      if (iter->second == 0xff) {
```
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:101]: Creating a variable for this `0xff` value at the top level scope would make this easier to understand:

```
constexpr uint8_t UnmatchedResp = 0xff;
```

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+  // iterate through all response dequeues to find out which ones haven't been consumed
+  for (auto& [packet_id, resp_deque] : *resp_frames) {
+    for (auto& resp : resp_deque) {
+      if (!resp.consumed) {
+        error_count++;
+      }
+    }
+    resp_deque.clear();
+  }
```
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:167]: For other protocols, we assume that a request will always be seen before a response. This allows us the ability to clear the response deque at the end of each stitching round. I'm not sure that assumption will hold for MQTT since the server and client can independently initiate `PUBLISH` requests (as I mentioned in another comment).

Can we add a comment here to make sure we test how server "client" frames work with this logic?

**Review by ddelnano — COMMENTED** [review]
@ChinmayaSharma-hue sorry for the late review and really appreciate all your work on the MQTT protocol support!

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+    // track the latest response timestamp to compare against request frame's timestamp later.
+    uint64_t latest_resp_ts = resp_deque.empty() ? 0 : resp_deque.back().timestamp_ns;
+    // finding the closest appropriate response from response deque in terms of timestamp and type
+    // for each request in the request deque
+    for (mqtt::Message& req_frame : req_deque) {
+      // getting the appropriate response match value for the request match key
+      uint8_t request_match_key =
+          (req_frame.control_packet_type << 4) | static_cast<uint8_t>(req_frame.dup) << 3 |
+          (req_frame.header_fields["qos"]) << 1 | static_cast<uint8_t>(req_frame.retain);
```
[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:94]: You mean something like this, right?
```
MatchKey getMatchKey(mqtt::Message frame) {
    return (frame.control_packet_type << 4) | static_cast<uint8_t>(frame.dup) << 3 |
    (frame.header_fields["qos"] & 0x3) << 1 | static_cast<uint8_t>(frame.retain);
}
```

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+                                    req_frame.timestamp_ns);
+        continue;
+      }
+
+      // finding the first appropriate response frame with the desired control packet type and flags
+      auto response_frame_iter = std::find_if(
+          first_timestamp_iter, resp_deque.end(), [response_match_value](mqtt::Message& message) {
+            return ((message.control_packet_type << 4) | (static_cast<uint8_t>(message.dup) << 3) |
+                    (message.header_fields["qos"] << 1) | static_cast<uint8_t>(message.retain)) ==
+                   response_match_value;
```
[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:128]: This would be changed to this according to the function for this logic,
```
auto response_frame_iter = std::find_if(
          first_timestamp_iter, resp_deque.end(), [response_match_value](mqtt::Message& message) {
            return getMatchKey(message) == response_match_value;
          });
```

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    std::deque<mqtt::Message>& resp_deque = (pos != resp_frames->end()) ? pos->second : empty_deque;
+
+    // track the latest response timestamp to compare against request frame's timestamp later.
+    uint64_t latest_resp_ts = resp_deque.empty() ? 0 : resp_deque.back().timestamp_ns;
+    // finding the closest appropriate response from response deque in terms of timestamp and type
+    // for each request in the request deque
+    for (mqtt::Message& req_frame : req_deque) {
+      // getting the appropriate response match value for the request match key
+      uint8_t request_match_key =
+          (req_frame.control_packet_type << 4) | static_cast<uint8_t>(req_frame.dup) << 3 |
```
[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:93]: > Can you add a comment [here](https://github.com/pixie-io/pixie/blob/7359d40df6272d15b73535df67928e760269390d/src/stirling/source_connectors/socket_tracer/protocols/mqtt/parse.cc#L40) to mention that this is modeling a 4 bit field?

> I also think we should cap the req_frame.header_fields["qos"] to 2 bits just to be safe. 

Sure, will do.
Should I also add a comment [here](https://github.com/pixie-io/pixie/blob/7359d40df6272d15b73535df67928e760269390d/src/stirling/source_connectors/socket_tracer/protocols/mqtt/types.h#L41) in types.h where `control_packet_type` is defined as `uint8_t` that it is modelling a 4 bit field?

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc [hunk 7]
```diff
@@ -0,0 +1,827 @@
...
+  req_map[0].push_back(connect_frame);
+  req_map[0].push_back(disconnect_frame);
+  req_map[0].push_back(auth_frame);
+  resp_map[0].push_back(connack_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 4);
+}
+
```
[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:823]: > Duplicate frame during a QoS 1 PUBLISH

I am assuming you mean a duplicate QOS 1 PUBLISH from the client itself, with a single PUBACK from the server (caused due to retransmission of PUBLISH after not receiving PUBACK). This pointed to a bug in the stitcher code where I was not checking if the response had been consumed in the response deque in the code where I was searching for a response with the matching response key. This was causing a single response to be matched to multiple duplicate requests. I have added this, and now the code works like this in case of duplicate QOS 1 PUBLISH,
1. If there are two duplicate PUBLISH frames (QOS 1) in the request deque, and one PUBACK frame in the response deque, then the first PUBLISH frame is stitched with the PUBACK frame, and the second PUBLISH frame is left in the request deque (with error count as 0). The code works like this because I did not want to clear certain request frames in the end of the deque, as the response could be filled the next time StitchFrames is called and the response deque now contains the response to this request. We know however that no response will come for this duplicate PUBLISH request, but I see no problem with this as this request will be cleared in the next round when more request-response stitching occurs in the deque.
2. If there are two duplicate QOS 1 PUBLISH frames, one SUBSCRIBE frame (can be any other request frame) in the request deque, and one PUBACK frame, one SUBACK frame in the response deque, then the first PUBLISH will be stitched with PUBACK, the SUBSCRIBE will be stitched with SUBACK, and this time the error count will be 1, and the duplicate PUBLISH will be cleared from the request deque.

> Case were client and server are using the same packet identifiers -- this is possible according to the spec and is different from how other protocols that use the StitchFrames map interface work (ID generation is usually client side only).

This would work fine if the flip-flopping as you mentioned does not occur, but if it does, then I will have to add logic to do reverse stitching from response to request as well, as you said. I will add a TODO comment for reference in the future.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc [hunk 8]
```diff
@@ -0,0 +1,827 @@
...
+TEST(MqttStitcherTest, OnlyRequests) {
+  absl::flat_hash_map<packet_id_t, std::deque<Message>> req_map;
+  absl::flat_hash_map<packet_id_t, std::deque<Message>> resp_map;
+
+  std::string_view frame_view;
+  RecordsWithErrorCount<Record> result;
+
+  Message connect_frame, pingreq_frame;
+  frame_view = CreateStringView<char>(CharArrayStringView<uint8_t>(kConnect));
+  ParseFrame(message_type_t::kRequest, &frame_view, &connect_frame);
```
[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:236]: I should have done this as well. I have made a CreateFrames function now that takes in type (Request/Response), control_packet_type, packet_identifier and qos as arguments,  and replaced all the code where I was parsing frames.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+#include "src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.h"
+
+#include <algorithm>
+#include <deque>
+#include <map>
+#include <string>
+#include <utility>
+
+#include "src/common/base/base.h"
+#include "src/common/json/json.h"
```
[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:28]: Removed.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc (hunk 2 above)
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:101]: Done.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc (hunk 3 above)
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:167]: This would be a follow up for the TODO comment for the case where PUBLISH ends up in the response deque, right? Another TODO comment to make sure we test the clearing of the response deque when such a thing occurs?

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:93]: Thanks for calling that out. Yep, that would be great to have as well.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:94]: Yep, that's what I had in mind. We should also avoid copying the `mqtt::Message` when this function is used as well.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:128]: Sounds good to me :+1:

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:823]: For case 1, how does the implementation discern between an unanswered `PUBLISH` frame (QOS 1) compared to a `PUBLISH` sequence (QOS1) where there are duplicated `PUBLISH`es and a `PUBACK`? It seems that the next round of stitching might have difficulty telling the two cases apart. Imo the former should be classified as an error while the latter seems fine to consider as a successful case.

Sounds good on the TODO and revisiting this for case 2.

[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:823]: > how does the implementation discern between an unanswered PUBLISH frame (QOS 1) compared to a PUBLISH sequence (QOS1) where there are duplicated PUBLISHes and a PUBACK? 

Actually the implementation would consider one of the duplicated PUBLISH frames (QOS 1) (or multiple, if there are more than one duplicate frames due to retransmission) as unanswered PUBLISH frames. In the next round of stitching, the unanswered PUBLISH frames would be discarded, when newer frames which have newer responses are stitched.
That logic is present in this code,
```
while (iter != req_deque.end() && (iter->timestamp_ns < latest_resp_ts)) {
  if (iter->consumed) {
    ++erase_until_iter;
  }
  if (!iter->consumed && !(iter == req_deque.end() - 1) && ((erase_until_iter + 1)->consumed)) {
    ++error_count;
    ++erase_until_iter;
  }
  ++iter;
}
req_deque.erase(req_deque.begin(), erase_until_iter);
}
```
where all the requests that have not been consumed until the latest response timestamp would be removed. I thought this would make sense, since if there are responses to subsequent requests after the unanswered PUBLISH, then it will go unanswered and I need to remove it from the request deque.

But yeah, it will give error count as 1 in cases where are there duplicate PUBLISH frames with a PUBACK. I thought this was fine since it is a case of an unanswered PUBLISH itself that prompted retransmission and duplication, so error count being the number of duplicate PUBLISH frames minus one (for the answered PUBLISH) would make sense.

Should I modify the logic so that the error count is zero in cases where the PUBLISH frames finally has an answer in PUBACK, and error count is non  zero in cases where there are no PUBACKs at all?

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc (hunk 3 above)
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:167]: Yeah that comment makes things clear.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:823]: > Should I modify the logic so that the error count is zero in cases where the PUBLISH frames finally has an answer in PUBACK, and error count is non zero in cases where there are no PUBACKs at all?

Yes, I believe that's the correct thing to do here. Stirling will disable tracing a connection if the stitcher failure rate goes over a certain threshold, so accurately labeling these cases will avoid difficult to debug situations later on.

[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:823]: I was thinking, the way I could do this is marking the duplicate messages that have been answered as consumed.

Also, the stitching logic would match the PUBLISH with `dup` bit set to 0 with the PUBACK, and the PUBLISH with `dup` bit set to 1 would be dropped. Is this fine? Should I replace the PUBLISH in the entries vector once a duplicate has been found that has already been stitched with a PUBACK with the first PUBLISH frame?

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc [hunk 10]
```diff
@@ -0,0 +1,623 @@
...
+  resp_map[0].push_back(connack_frame);
+  req_map[0].push_back(pingreq_frame);
+  resp_map[0].push_back(pingresp_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_TRUE(AreAllDequesEmpty(resp_map));
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 2);
+  req_map.clear();
+  resp_map.clear();
```
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:283]: Previous assertions prove that the maps are cleared, please remove.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc [hunk 11]
```diff
@@ -0,0 +1,623 @@
...
+  resp_map[1].push_back(suback_frame);
+  resp_map[1].push_back(puback_frame);
+  resp_map[2].push_back(unsuback_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_TRUE(AreAllDequesEmpty(resp_map));
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 3);
+  req_map.clear();
+  resp_map.clear();
```
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:304]: Previous assertions prove that the maps are cleared, please remove.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc [hunk 12]
```diff
@@ -0,0 +1,623 @@
...
+  resp_map[1].push_back(pubrec_frame);
+  resp_map[1].push_back(pubcomp_frame);
+  resp_map[2].push_back(unsuback_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_TRUE(AreAllDequesEmpty(resp_map));
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 4);
+  req_map.clear();
+  resp_map.clear();
```
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:329]: Previous assertions prove that the maps are cleared, please remove.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc [hunk 13]
```diff
@@ -0,0 +1,623 @@
...
+  connack_frame.timestamp_ns = ++t;
+  req_map[0].push_back(connect_frame);
+  resp_map[0].push_back(connack_frame);
+  req_map[0].push_back(pingreq_frame);
+  resp_map[0].push_back(pingresp_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_TRUE(AreAllDequesEmpty(resp_map));
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 2);
```
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:378]: This is already present in first assertions of `InOrderMatching`. I think we can remove to focus on the test coverage that is unique to this test.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc [hunk 14]
```diff
@@ -0,0 +1,623 @@
...
+  req_map[0].push_back(connect_frame);
+  req_map[0].push_back(pingreq_frame);
+  resp_map[0].push_back(pingresp_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_TRUE(AreAllDequesEmpty(resp_map));
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 1);
+  EXPECT_EQ(result.records.size(), 1);
+  req_map.clear();
+  resp_map.clear();
```
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:134]: Previous assertions prove map is cleared, please remove these `.clear()` calls.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc [hunk 15]
```diff
@@ -0,0 +1,623 @@
...
+  req_map[2].push_back(unsub_frame);
+  resp_map[1].push_back(puback_frame);
+  resp_map[2].push_back(unsuback_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_TRUE(AreAllDequesEmpty(resp_map));
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 1);
+  EXPECT_EQ(result.records.size(), 2);
+  req_map.clear();
+  resp_map.clear();
```
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:154]: Please remove `.clear()` calls since previous assertions show maps are empty.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:477]: > Also, the stitching logic would match the PUBLISH with dup bit set to 0 with the PUBACK, and the PUBLISH with dup bit set to 1 would be dropped. Is this fine?

It feels like the response (PUBREC, etc) should ideally be matched against the latest dup. Matching against the non duplicate frame gives the false impression that the duplicate shouldn't be expected (meaning the original message was delivered properly).

Maybe we can revisit this thread after you have a chance to consider my comments about using state to handle the duplicates.

**Review by ddelnano — COMMENTED** [review]
Thanks for the latest updates! I know this latest round has quite a bit of comments, but stitching is a challenging component and the more thought that's put into the edge cases now will save time when we move onto the next step!

As for the tests that clear the reqs/resps maps before the next batch of assertions, it might be helpful to split each of those into their own test case (`TEST(MqttStitchTest, XXX)` macro). I assume the map `.clears()` are there to prevent the earlier "test cases" from causing unrelated failures for the later test code. In my opinion, that is a signal that they should be independent cases and any shared setup should be refactored into another function.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+    // UNSUBSCRIBE to UNSUBACK
+    {0xa0, 0xb0},
+    // PINGREQ to PINGRESP
+    {0xc0, 0xd0},
+    // DISCONNECT to Dummy response
+    {0xe0, UnmatchedResp},
+    // AUTH to Dummy response
+    {0xf0, UnmatchedResp}};
+
+std::set<std::tuple<uint32_t, uint32_t>> UnansweredPublish;
```
[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:72]: I was wondering if when the server initiates PUBLISH, the message type is going to be a `kRequest` or a `kResponse`. I am assuming that it will be a `kResponse`, since in `conn_tracker.cc` and `conn_tracker.h` it looks like whether the type is passed as `kRequest` or `kResponse` depends on the role of the sender,
conn_tracker.cc,
```cpp
DataStream* ConnTracker::req_data() {
  DCHECK_NE(role_, kRoleUnknown);
  switch (role_) {
    case kRoleClient:
      return &send_data_;
    case kRoleServer:
      return &recv_data_;
    default:
      return nullptr;
  }
}
```
conn_tracker.h,
```cpp    
req_data_ptr->template ProcessBytesToFrames<TKey, TFrameType, TStateType>(
    message_type_t::kRequest, state_ptr);

DataStream* resp_data_ptr = resp_data();
DCHECK_NE(resp_data_ptr, nullptr);
resp_data_ptr->template ProcessBytesToFrames<TKey, TFrameType, TStateType>(
    message_type_t::kResponse, state_ptr);
```
This would help me put increment the counter on the right map in the state.

Other than that, this idea seems like a great solution for the duplicate PUBLISH problem. Thanks for pointing it out!

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc [hunk 17]
```diff
@@ -0,0 +1,623 @@
...
+  // Test for requests that do not have responses
+  absl::flat_hash_map<packet_id_t, std::deque<Message>> req_map;
+  absl::flat_hash_map<packet_id_t, std::deque<Message>> resp_map;
+
+  RecordsWithErrorCount<Record> result;
+
+  Message pub0_frame, disconnect_frame, auth_frame, connect_frame, connack_frame;
+  pub0_frame = CreateFrame(kRequest, 3, 0, 0);         // PUBLISH with QoS 0
+  disconnect_frame = CreateFrame(kRequest, 14, 0, 0);  // DISCONNECT
+  auth_frame = CreateFrame(kRequest, 15, 0, 0);        // AUTH
```
[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:443]: This actually complicates things. This would be similar to the case of the request-response being in the opposite direction, which would put the first AUTH packet sent by the server in the response deque, and the AUTH response by the client in the request deque. I have the logic of checking for packets with timestamps greater than the request frame in the response deque, which the server sent AUTH request wouldn't have.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc [hunk 18]
```diff
@@ -0,0 +1,623 @@
...
+  connack_frame.timestamp_ns = ++t;
+  req_map[0].push_back(pub0_frame);
+  req_map[0].push_back(connect_frame);
+  req_map[0].push_back(disconnect_frame);
+  req_map[0].push_back(auth_frame);
+  resp_map[0].push_back(connack_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 4);
```
[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:475]: Yeah sure! I will make sure to remove redundant test cases.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc [hunk 19]
```diff
@@ -0,0 +1,623 @@
...
+  req_map[1].push_back(pub1_frame_2);
+  req_map[1].push_back(subscribe_frame);
+  resp_map[1].push_back(puback_frame);
+  resp_map[1].push_back(suback_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 2);
+  req_map.clear();
+  resp_map.clear();
```
[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:528]: Yeah sure! I will make sure to remove redundant test cases.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc [hunk 20]
```diff
@@ -0,0 +1,623 @@
...
+  req_map[1].push_back(pub2_frame_2);
+  req_map[1].push_back(subscribe_frame);
+  resp_map[1].push_back(pubrec_frame);
+  resp_map[1].push_back(suback_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 2);
+  req_map.clear();
+  resp_map.clear();
```
[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:579]: Yeah sure! I will make sure to remove redundant test cases.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc (hunk 10 above)
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:283]: Yeah sure! I will make sure to remove redundant test cases.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc (hunk 13 above)
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:378]: Yeah sure! I will make sure to remove redundant test cases.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:477]: Yeah using state to handle duplicates makes this really simple, now the latest duplicate can be matched to the response.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc (hunk 16 above)
[the remark this answers is not in the document]
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:72]: I believe it's going to be a `kResponse` as well. Since this is the first protocol that Pixie will support that is bidirectional (client and server can send requests), we will need to verify that -- which is why I wanted to leave a TODO (from [this thread](https://github.com/pixie-io/pixie/pull/1918#discussion_r1643096915)). Our common heuristics (clearing all unmatched response frames) assume that the server won't send requests (labeled as responses).

No problem!

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (ddelnano) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:443]: How important do you think it is to capture these AUTH frames? From my perspective, I think supporting these frames isn't the most critical for the initial support. I'd be ok with consuming / dropping them and leaving a TODO comment for now with a brief explanation of the challenges.

[the remark this answers is not in the document]
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc:443]: Yeah I think we can skip doing stitching for Auth frames for now. The only time it is sent by the server is when authentication is not complete in the connection stage itself (I think). We can support them later.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/parse.cc [hunk 21]
```diff
@@ -724,6 +706,27 @@ ParseState ParseFrame(message_type_t type, std::string_view* buf, Message* resul
...
+      }
+    } else {
+      if (type == message_type_t::kRequest) {
+        state->send[std::tuple<uint32_t, uint32_t>(result->header_fields["packet_identifier"],
+                                                   result->header_fields["qos"])] = 0;
+      } else {
+        state->recv[std::tuple<uint32_t, uint32_t>(result->header_fields["packet_identifier"],
+                                                   result->header_fields["qos"])] = 0;
+      }
+    }
```
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/protocols/mqtt/parse.cc:727]: Is it necessary to store a key in the state for `PUBLISH`'es that don't have any dups?

The `::operator[]` method will insert into the map if the key doesn't exist. So I believe the `+=` operator will work as expected without the non dup branch.

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+        if (req_frame.type == message_type_t::kRequest &&
+            state->send[unique_publish_identifier] > 0) {
+          state->send[unique_publish_identifier] -= 1;
+          req_frame.consumed = true;
+          continue;
+        }
+
+        if (req_frame.type == message_type_t::kResponse &&
+            state->recv[unique_publish_identifier] > 0) {
+          state->recv[unique_publish_identifier] -= 1;
```
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:123]: Related to my earlier comment about avoiding excess inserts into `state->{recv,send}`. Can we use `::find()` instead of `::operator[]`?

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+    {0x3d, 0x50},
+    // PUBREL to PUBCOMP
+    {0x60, 0x70},
+    // SUBSCRIBE to SUBACK
+    {0X80, 0X90},
+    // UNSUBSCRIBE to UNSUBACK
+    {0xa0, 0xb0},
+    // PINGREQ to PINGRESP
+    {0xc0, 0xd0},
+    // AUTH to AUTH
```
Comment (ddelnano) [inline src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:68]: I thought we decided that we were skipping supporting `AUTH` frames for now. If that changed, I'm ok with keeping this. However, I'd like to add a comment explaining what the gap is (if any) while it's still fresh in our minds.

**Review by ddelnano — COMMENTED** [review]
Thanks again for all the hard work on this. This round of feedback is only a few minor things and I expect that after these are addressed we can move forward with merging this!

File: src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc (hunk 23 above)
Comment (ChinmayaSharma-hue) [reply src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher.cc:68]: I decided to remove the mapping from the map itself, and a TODO before the logic of skipping AUTH matching in the stitcher function.

**Review by ddelnano — APPROVED** [review]
